tableaudocumentapi/datasource.py

```diff
@@ -258,3 +259,38 @@ def _get_metadata_objects(self):
     def _get_column_objects(self):
         return [_column_object_from_column_xml(self._datasourceTree, xml)
                 for xml in self._datasourceTree.findall('.//column')]
+
+    ###########
+    # folders
+    ###########
+    @property
+    def folders(self):
+        if not self._folders:
+            self._refresh_folders()
+        return self._folders
+
+    def _refresh_folders(self):
+        folders = Folder.all_folders_from_datasource(self)
+        self._folders = {f.name: f for f in folders}
+
+    def add_folder(self, name, role):
+        """ Adds a new, empty folder to the datasource and returns it.
+
+        Will fail if another folder of the same name already exists,
+        or if no valid role (dimensions or measures) is provided.
+        """
+
+        if name in self.folders.keys():
+            raise ValueError('Folder names must be unique')
+
+        folder = Folder.from_name_and_role(self, name, role)
+        # find the index of the last column to insert the folder
```

Comment:
Thank you! It does!

The good news, is I don't think we need to support that. It's an implementation detail that all folder elements happen to come after the column elements, and I'm not confidant it's intentional or won't change -- as long as Desktop can open a file where you don't bother to do this, I think we can save some code!

It also means the library won't be as married to the XML format, should we ever update our document format.